modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,698 @@
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI input are not set");
+            }
+            if (std::any_of(el.outputs.cbegin(), el.outputs.cend(),
+                            [](ExtractTypeHelper::OutputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI outputs are not set");
+            }
+        }
+
+        // Prepare nodes for linking
+        std::vector<ade::NodeHandle> in_nodes;
+        std::vector<ade::NodeHandle> out_nodes;
+        std::vector<ade::NodeHandle> inter_nodes;
+
+        for (const auto& node : m_oak_nodes) {
+            auto nh = node.first;
+            // Fill input op nodes
+            for (const auto& d : ins_data) {
+                for (const auto& indata : nh.get()->inNodes()) {
+                    auto rc = m_gm.metadata(indata).get<cv::gimpl::Data>().rc;
+                    if (rc == d.rc) {
+                        in_nodes.push_back(nh);
+                    }
+                }
+            }
+            // Fill output op nodes
+            for (const auto& d : outs_data) {
+                for (const auto& outdata : nh.get()->outNodes()) {
+                    auto rc = m_gm.metadata(outdata).get<cv::gimpl::Data>().rc;
+                    if (rc == d.rc) {
+                        out_nodes.push_back(nh);
+                    }
+                }
+            }
+            // Fill internal op nodes
+            if (std::find(in_nodes.begin(), in_nodes.end(), nh) == in_nodes.end() &&
+                std::find(out_nodes.begin(), out_nodes.end(), nh) == out_nodes.end()) {
+                inter_nodes.push_back(nh);
+            }
+        }
+
+        // Properly link all nodes
+        // 1. Link input nodes to camera
+        for (const auto& nh : in_nodes) {
+            GAPI_Assert(m_oak_nodes.at(nh).inputs.size() == 1);
+            // FIXME: covert other camera outputs
+            m_camera_input->video.link(*(m_oak_nodes.at(nh).inputs[0]));
+        }
+
+        // 2. Link output nodes to XLinkOut nodes
+        size_t out_counter = 0;
+        for (const auto& nh : out_nodes) {
+            GAPI_Assert(m_oak_nodes.at(nh).outputs.size() == 1);
+            GAPI_Assert(out_counter < m_xlink_outputs.size());
+            m_oak_nodes.at(nh).outputs[0]->link(m_xlink_outputs[out_counter++]->input);
+            // Input nodes in OAK doesn't have parent operation - just camera (for now)
+            if (std::find(in_nodes.begin(), in_nodes.end(), nh) == in_nodes.end()) {
+                LinkToParents(nh);
+            }
+        }
+
+        // 3. Link internal nodes to their parents
+        for (const auto& nh : inter_nodes) {
+            // Input nodes in OAK doesn't have parent operation - just camera (for now)
+            if (std::find(in_nodes.begin(), in_nodes.end(), nh) == in_nodes.end()) {
```

Comment:
This search is linear too. Also why do you need this check here at all? Is your `inter_nodes` incorrect?

Reply:
Done